### PR TITLE
FocusWindow command

### PR DESCRIPTION
**Pull request by Melesar**
This command focuses a window with the given `WM_CLASS`. 

- If the window is not present, it does nothing
- If the window is visible in any workspace, it will gain focus
- If the window is not visible, the current workspace will switch to a tag with the window and give it focus
- If the window is hidden in the `Monocle` or `MainAndDeck` layout, it will scroll the layout so that the window becomes visible and focused
- If the window is already focused, it will focus the previousely focused one, following the same rules

Example configuration:

```toml
[[keybind]]
command = "FocusWindow"
value = "brave-browser" # The first or the second entry of WM_CLASS attribute
modifier = ["modkey"]
key = "b"
```
Another added possibility is to focus a visible window by its index on the screen. Example configuration:

```toml
[[keybind]]
command = "FocusWindow"
value = "1" # Will focus the first window on the screen
modifier = ["modkey", "Control"]
key = "1"
```
Indices are not displayed in any way yet. This feature is WIP for now and will not be documented

### CODE REVIEW

**Comment by AethanFoot** [comment]
Ill have a proper look on my pc tomorrow, it may be nice if we could expand this to accept other tags, e.g. window id, window name, etc. Some these wont be as useful straight away, but would be useful in the future when users may query leftwm for say a set of window ids which they can implement into a script nicely. 

With msrv we may need to update the clippy or bring it up (depending when that was added as we don't want to break for slower updating distros) @mautamu what do you think?

**Comment by mautamu** [comment]
I'm good with moving up the MSRV personally. However, we'd need to move it to a 2021-compatible stable compiler (>=1.56.0). Seeing as to how 1.59.0 should be coming out in three weeks, moving to 1.56.0 sounds reasonable to me.

Tagging https://github.com/leftwm/leftwm/pull/594 @hertg

**Comment by mautamu** [comment]
Would someone with write access to this PR/branch test a0192f8 with https://github.com/leftwm/leftwm/blob/a0192f8c59c32f0fdb2bf28fd451c6379a369943/.github/workflows/ci.yml#L54 changed to read
```yaml
    toolchain: 1.56.1
```?

**Comment by VuiMuich** [comment]
> Would someone with write access to this PR/branch test [a0192f8](https://github.com/leftwm/leftwm/commit/a0192f8c59c32f0fdb2bf28fd451c6379a369943) with
> 
> https://github.com/leftwm/leftwm/blob/a0192f8c59c32f0fdb2bf28fd451c6379a369943/.github/workflows/ci.yml#L54
> 
> changed to read
> ```yaml
>     toolchain: 1.56.1
> ```
> 
> ?

Right now all I can offer is:
```
on  go_to_window_cmd [$] via 🦀 v1.58.1
leftwm >>> make test
cd /home/vuimuich/Git/leftwm && cargo test --all-targets --all-features
   Compiling leftwm-core v0.3.0 (/home/vuimuich/Git/leftwm/leftwm-core)
   Compiling leftwm v0.3.0 (/home/vuimuich/Git/leftwm/leftwm)
    Finished test [unoptimized + debuginfo] target(s) in 24.34s
     Running unittests (target/debug/deps/leftwm-43e6c1f29c172fc1)

running 1 test
test theme_setting::tests::deserialize_custom_theme_config ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests (target/debug/deps/leftwm-2a0f93ac591f6e9e)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests (target/debug/deps/leftwm_check-4be31c870a388d55)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests (target/debug/deps/leftwm_command-8f4d7b01c182384f)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests (target/debug/deps/leftwm_state-6ae354b636c93455)

running 1 test
test tests::correctly_identifies_partial_template_filenames ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests (target/debug/deps/leftwm_worker-0a4bdf897bd02c92)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests (target/debug/deps/leftwm_core-76593bfac7b5d470)

running 73 tests
test config::tests::ensure_command_handler_trait_boundary ... ok
test config::tests::check_wm_class_is_associated_with_predefined_tag ... ok
test handlers::command_handler::tests::focus_tag_change_should_go_to_previous_and_next_tag ... ok
test handlers::command_handler::tests::go_to_tag_should_return_false_if_no_screen_is_created ... ok
test handlers::command_handler::tests::go_to_tag_should_create_at_least_one_tag_per_screen_no_more ... ok
test handlers::command_handler::tests::focus_window_top ... ok
test handlers::focus_handler::tests::focusing_a_tag_should_focus_its_workspace ... ok
test handlers::command_handler::tests::go_to_tag_should_go_to_tag_and_set_history ... ok
test handlers::focus_handler::tests::focusing_a_tag_should_make_it_active ... ok
test handlers::command_handler::tests::go_to_tag_should_return_false_on_invalid_input ... ok
test handlers::focus_handler::tests::focusing_a_window_should_focus_workspace ... ok
test handlers::focus_handler::tests::focusing_a_window_should_focus_its_tag ... ok
test handlers::focus_handler::tests::focusing_a_window_should_make_it_active ... ok
test handlers::focus_handler::tests::focusing_a_workspace_should_focus_its_last_active_window ... ok
test handlers::focus_handler::tests::focusing_a_workspace_should_make_it_active ... ok
test handlers::focus_handler::tests::focusing_the_same_tag_shouldnt_add_to_the_history ... ok
test handlers::focus_handler::tests::focusing_a_workspace_should_focus_its_tag ... ok
test handlers::focus_handler::tests::focusing_an_empty_tag_should_unfocus_any_focused_window ... ok
test handlers::focus_handler::tests::focusing_the_same_window_shouldnt_add_to_the_history ... ok
test handlers::focus_handler::tests::focusing_the_same_workspace_shouldnt_add_to_the_history ... ok
test handlers::goto_tag_handler::tests::going_to_a_workspace_that_is_already_visible_should_not_duplicate_the_workspace ... ok
test handlers::screen_create_handler::tests::creating_two_screens_should_tag_them_with_first_and_second_tags ... ok
test layouts::tests::should_fullscreen_a_single_window ... ok
test layouts::tests::test_from_str ... ok
test handlers::screen_create_handler::tests::should_be_able_to_add_screens_with_preexisting_tags ... ok
test models::dock_area::tests::should_be_able_to_build_from_left ... ok
test models::dock_area::tests::should_be_able_to_build_from_right ... ok
test models::dock_area::tests::should_be_able_to_build_from_top ... ok
test models::layout_manager::tests::layouts_should_fallback_to_the_global_list ... ok
test models::layout_manager::tests::next_layout_basic ... ok
test models::layout_manager::tests::next_layout_fallback_to_global_layouts ... ok
test models::layout_manager::tests::next_layout_fallback_to_the_first_layout ... ok
test handlers::screen_create_handler::tests::creating_more_screens_than_tags_should_automatically_create_new_tags ... ok
test models::layout_manager::tests::next_layout_should_cycle ... ok
test models::layout_manager::tests::prev_layout_basic ... ok
test models::dock_area::tests::should_be_able_to_build_from_bottom ... ok
test models::layout_manager::tests::prev_layout_should_cycle ... ok
test models::layout_manager::tests::previous_layout_fallback_to_global_layouts ... ok
test models::layout_manager::tests::previous_layout_fallback_to_the_first_layout ... ok
test models::tag::tests::hidden_tags_must_be_retrievable_by_label ... ok
test models::tag::tests::hidden_tags_must_have_unique_label ... ok
test models::tag::tests::multiple_normal_tags_can_have_same_label ... ok
test models::tag::tests::must_be_able_to_get_all_tags ... ok
test models::tag::tests::must_be_able_to_only_get_normal_tags ... ok
test models::tag::tests::normal_tags_are_numbered_in_order ... ok
test models::tag::tests::tags_can_be_mutable ... ok
test models::tag::tests::unlabelled_tags_are_automatically_labelled_with_their_id ... ok
test models::window::tests::should_be_able_to_tag_a_window ... ok
test models::tag::tests::only_hidden_tags_can_be_retrieved_by_label ... ok
test models::tag::tests::hidden_tags_are_numbered_in_order ... ok
test models::window::tests::should_be_able_to_untag_a_window ... ok
test models::workspace::tests::empty_ws_should_not_contain_window ... ok
test models::workspace::tests::tagging_a_workspace_to_with_the_same_tag_as_a_window_should_couse_it_to_display ... ok
test models::xyhw::tests::center_halfed ... ok
test models::xyhw::tests::contains_xyhw_should_detect_a_inner_window ... ok
test models::xyhw::tests::contains_xyhw_should_detect_a_lower_right_corner_outside ... ok
test models::xyhw::tests::contains_xyhw_should_detect_a_upper_left_corner_outside ... ok
test models::xyhw::tests::without_should_trim_from_the_bottom ... ok
test models::xyhw::tests::without_should_trim_from_the_left ... ok
test models::xyhw::tests::without_should_trim_from_the_right ... ok
test models::xyhw::tests::without_should_trim_from_the_top ... ok
test utils::child_process::tests::test_parse ... ok
test utils::helpers::test::relative_find_loop_can_be_disabled ... ok
test utils::helpers::test::relative_find_should_be_able_to_loop ... ok
test utils::helpers::test::relative_find_should_work_both_ways ... ok
test utils::helpers::test::relative_find_with_inexistent_reference_must_return_none ... ok
test utils::command_pipe::test::read_good_command ... ok
test utils::command_pipe::test::read_bad_command ... ok
test utils::state_socket::test::get_update ... ok
test utils::state_socket::test::multiple_peers ... ok
test utils::state_socket::test::socket_already_bound ... ok
test utils::state_socket::test::socket_cleanup ... ok
test utils::command_pipe::test::pipe_cleanup ... ok

test result: ok. 73 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.12s

cd /home/vuimuich/Git/leftwm && cargo fmt -- --check
cd /home/vuimuich/Git/leftwm && cargo clippy --release
    Checking leftwm-core v0.3.0 (/home/vuimuich/Git/leftwm/leftwm-core)
warning: unnested or-patterns
   --> leftwm-core/src/handlers/command_handler.rs:308:9
    |
308 |         Some(Layout::Monocle) | Some(Layout::MainAndDeck) => {
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
note: the lint level is defined here
   --> leftwm-core/src/lib.rs:5:9
    |
5   | #![warn(clippy::pedantic)]
    |         ^^^^^^^^^^^^^^^^
    = note: `#[warn(clippy::unnested_or_patterns)]` implied by `#[warn(clippy::pedantic)]`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnested_or_patterns
help: nest the patterns
    |
308 |         Some(Layout::Monocle | Layout::MainAndDeck) => {
    |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

warning: `leftwm-core` (lib) generated 1 warning
    Checking leftwm v0.3.0 (/home/vuimuich/Git/leftwm/leftwm)
    Finished release [optimized] target(s) in 7.44s
```
More possibly later today.

**Comment by Melesar** [comment]
I wasn't sure about that msrv check, so I changed `Some(Layout::Monocle | Layout::MainAndDeck)` to `Some(Layout::Monocle) | Some(Layout::MainAndDeck)`. So I guess I'll revert that to make clippy happy again :)

File: leftwm-core/src/handlers/command_handler.rs
Comment (AethanFoot) [inline leftwm-core/src/handlers/command_handler.rs:316]: I would check whether the window in floating first/whether it is visible again. You will want to exclude floating windows from the extraction as we don't want to cycle through these, the same for the main window in main and deck as this should be fixed.

**Review by AethanFoot — COMMENTED** [review]
Nice addition!

File: leftwm-core/src/handlers/command_handler.rs
Comment (Melesar) [reply leftwm-core/src/handlers/command_handler.rs:282]: Not really. In fact, looking at some other examples in the codebase, I think it will be more consistent to also check `res_class`. Will do that

File: leftwm-core/src/handlers/command_handler.rs
Comment (AethanFoot) [inline leftwm-core/src/handlers/command_handler.rs:307]: This should return None now e.g.
```
focus(...);
return None;
```
As no windows need moving/resizing.

File: leftwm-core/src/handlers/command_handler.rs
Comment (AethanFoot) [inline leftwm-core/src/handlers/command_handler.rs:310]: After this point we should always return `Some(true)` as we have moved tags so windows need to be redrawn.

**Review by AethanFoot — CHANGES_REQUESTED** [review]
Other than bringing this inline with changes in #690, this looks good.

**Comment by AethanFoot** [comment]
Would you be okay adding part of #534, where you would focus the window by its index (don't worry about the overlay and such as these will be added later).

**Comment by Melesar** [comment]
Sure I can do that. Do you think it should be a separate command, or we can consolidate these two somehow?

Also I'm not sure how to bind it in the config file. I assume something like it's done with tags, i.e.
```toml
[[keybind]]
command = "JumpToWindow"
value = 1
modifier = ["modkey", "Control"]
key = "1"

[[keybind]]
command = "JumpToWindow"
value = 2
modifier = ["modkey", "Control"]
key = "2"

# e.t.c
```
As far as I know there is no such thing as dynamic keybinding yet

**Comment by VuiMuich** [comment]
I guess it would be nice to consolidate and make the distinction by the type of `value`: `integer` will jump like proposed in #534, a `string` will be parsed for `res_class`, `res_name` and at a later point even `window_id` or `pid`.

**Comment by Melesar** [comment]
> I guess it would be nice to consolidate and make the distinction by the type of `value`: `integer` will jump like proposed in #534, a `string` will be parsed for `res_class`, `res_name` and at a later point even `window_id` or `pid`.

Sounds good. However, without the overlay there will be usability issues with this mode, in my opinion ;)

**Comment by VuiMuich** [comment]
> > I guess it would be nice to consolidate and make the distinction by the type of `value`: `integer` will jump like proposed in #534, a `string` will be parsed for `res_class`, `res_name` and at a later point even `window_id` or `pid`.
> 
> Sounds good. However, without the overlay there will be usability issues with this mode, in my opinion ;)

Yeah, I believe that was the 'later' part Aethan referred to :grin:
So we might want to keep this a little bit under the hood (maybe add some comments in the code) for now and promote it properly when we are able to render the overlay, can do more fancy config stuff etc.

**Comment by Melesar** [comment]
Should work like this. I made the index 1-based, which is what the #534 TS seems to want and also this makes sense to me. Also the issue mentions **visible** windows, so there is no need for special treatment for Monocle and Main&Deck

File: leftwm/src/bin/leftwm-command.rs
Comment (AethanFoot) [inline leftwm/src/bin/leftwm-command.rs:90]: Other than updating this for the by index focus everything looks good!

**Review by AethanFoot — APPROVED** [review]

**Comment by Melesar** [comment]
Hi! Just wanted to clarify the workflow. 
Do we now wait for something/somebody to merge this? Or it's ready to be merged, it just will happen at a later time?

**Comment by VuiMuich** [comment]
Tbh we are a bit chaotic in that regard 🤷🏼‍♂️
Sometimes we gather PRs and merge a bunch at once sometimes we sporadically decide a PR is ready and merge. 

In this case I guess it would make sense to merge before #682 as it's probably easier bring your changes to it then the other way round.
What do you think @AethanFoot?

**Pull request merged**
